torchinductor/codegen/cpp.py

```diff
@@ -558,7 +910,35 @@ def codegen_nodes(self, nodes):
                     with kernel.write_to_suffix():
                         node.run(vars, ())
 
-        kernel_group.finalize_kernel(kernel, scheduler)
+        return kernel
+
+    def codegen_nodes(self, nodes):
+        """
+        Turn an set of pre-fused nodes into a C++ kernel.
+        """
+        kernel_group = self.kernel_group
+
+        can_be_simd_vec = self.check_simd_vec(nodes)
+        simd_vec_kernel = (
+            self._codegen_nodes_impl(nodes, can_be_simd_vec)
+            if can_be_simd_vec
+            else None
+        )
+        simd_omp_kernel = self._codegen_nodes_impl(nodes)
+
+        assert simd_omp_kernel
+        metrics.generated_kernel_count -= 1
+        # Maitain the metrics kernel count
+        if simd_vec_kernel:
+            metrics.generated_kernel_count -= 1
+
+        cpp_kernel_proxy = CppKernelProxy(
+            kernel_group.args, kernel_group.ws.num_threads
+        )
+        cpp_kernel_proxy.simd_vec_kernel = simd_vec_kernel
+        cpp_kernel_proxy.simd_omp_kernel = simd_omp_kernel
```

Comment:
Only one kind of kernel would be used at a time, right? Then, why can't we do the dispatch here directly to invoke either the original CppKernel or the vec Cppkernel? Why have to wrap another `CppKernelProxy` here?

Reply:
Nope. Actually, we only use the code of `CppKernel` and `CppVecKernel` but do not use its loops. The pseudo code is as follows.
```python
for cpp_kernel_proxy_loop_1:
  for cpp_kernel_proxy_loop_2:
    for main loop of cpp_kernel_proxy_loop_with_tail:
      CppVecKernel.load
      CppVecKernel.compute
      CppVecKernel.store
    for tail loop of cpp_kernel_proxy_loop_with_tail:
      CppKernel.load
      CppKernel.compute
      CppKernel.store
```